payara-micro-remote/src/main/java/fish/payara/arquillian/container/payaramicro/remote/DeployerClient.java

```diff
@@ -74,10 +73,20 @@
             httpConnection.setDoOutput(true);
             httpConnection.setDoInput(true);
             copy(archive, httpConnection.getOutputStream());
-            if (httpConnection.getResponseCode() == 200) {
+            int responseCode = httpConnection.getResponseCode();
+            if (responseCode == 200) {
                 return parseOutput(httpConnection.getInputStream());
             } else {
-                throw new IllegalArgumentException("Deployment failed. "+this+" returned "+httpConnection.getResponseCode());
+                JsonObject json = Json.createReader(httpConnection.getErrorStream()).readObject();
+                String message = json.getString("message", null);
+                if (message != null) {
+                    if (message.startsWith("CDI deployment failure") || message.contains("org.jboss.weld.exceptions.DeploymentException")) {
+                        throw new DeploymentException("Deployment failed. " + this + " returned " + responseCode);
+                    } else if (message.startsWith("CDI definition failure")) {
+                        throw new DefinitionException("Deployment failed. " + this + " returned " + responseCode);
+                    }
+                }
+                throw new IllegalArgumentException("Deployment failed. " + this + " returned " + responseCode);
```

Comment:
Let's not throw message away now that we captured it.

I'd propose throwing `DeploymentException` with a message whenever message is available. (and Definition exception in the special case).

I wouldn't worry about compatibility with `IllegalArgumentException` being thrown, as Arquillian doesn't really care about the type of exception thrown at deployment.

Reply:
I decided to stay with `IllegalArgumentException` as a general exception in case the deployment problem isn't related to CDI at all. So I'd rather keep it. But it's a good point to include the message in the exception message.